File: src/main/java/com/hiberus/anaya/redmineeditor/components/SummaryComponent.java
package com.hiberus.anaya.redmineeditor.components;

import com.hiberus.anaya.redmineeditor.utils.FXUtils;
import com.hiberus.anaya.redmineeditor.utils.hiberus.Schedule;
import javafx.fxml.FXML;
import javafx.scene.control.Label;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.FormatStyle;
import java.util.Set;

import static com.hiberus.anaya.redmineeditor.Model.ModelEditor.Events.*;
import static com.hiberus.anaya.redmineeditor.utils.TimeUtils.formatHours;

/**
 * A simple label with info of the current selected day
 */
public class SummaryComponent extends BaseComponent {

    /* ------------------------- views ------------------------- */

    @FXML
    private Label summary; // the label

    /* ------------------------- actions ------------------------- */


    @Override
    void init() {
        // when month, day, hours or the loading state changes, update
        controller.register(Set.of(Month, Day, Hours, Loading), model -> {
            if (model.isLoading()) {
                // while loading, notify user
                summary.setText("Loading...");
                summary.setBackground(null);
                return;
            }
            LocalDate date = model.getDate();
            if (date == null) {
                // if nothing selected, just ask
                summary.setText("Select day");
                summary.setBackground(null);
            } else {
                // on something selected, display info
                double spent = model.getSpent(date);
                double expected = Schedule.getExpectedHours(date);
                summary.setText(
                        date.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG))
                                + " --- Time: " + formatHours(spent) + "/" + formatHours(expected)
                                + (spent < expected ? " --- Missing: " + formatHours(expected - spent)
                                : spent > expected ? " --- Extra: " + formatHours(spent - expected)
                                : spent == expected && expected != 0 ? " --- OK"
                                : "")
                );

                // and change color
                FXUtils.setBackgroundColor(summary, Schedule.getColor(expected, spent, date));
            }
        });
    }
}

File: src/main/java/com/hiberus/anaya/redmineeditor/components/SummaryComponent.kt
package com.hiberus.anaya.redmineeditor.components

import com.hiberus.anaya.redmineeditor.controller.AppController
import com.hiberus.anaya.redmineeditor.model.ChangeEvents
import com.hiberus.anaya.redmineeditor.model.Model
import com.hiberus.anaya.redmineeditor.utils.backgroundColor
import com.hiberus.anaya.redmineeditor.utils.formatHours
import com.hiberus.anaya.redmineeditor.utils.hiberus.expectedHours
import com.hiberus.anaya.redmineeditor.utils.hiberus.getColor
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.paint.Color
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

/**
 * A simple label with info of the current selected day
 */
internal class SummaryComponent {

    /* ------------------------- views ------------------------- */

    @FXML
    lateinit var summary: Label // the label

    /* ------------------------- actions ------------------------- */

    @FXML
    fun initialize() {
        AppController.onChanges(
            // when month, day or hours changes, update
            setOf(ChangeEvents.Month, ChangeEvents.Day, ChangeEvents.DayHours, ChangeEvents.EntryList, ChangeEvents.Loading)
        ) updating@{ model: Model ->

            val date = model.date ?: run {
                // if nothing selected, just ask
                set("Select day")
                return@updating
            }
            val spent = model.getSpent(date) ?: run {
                // data not loaded yet, maybe loading
                set(if (model.isLoading) "Loading..." else "No data")
                return@updating
            }
            val expected = date.expectedHours

            // on something selected, display info with color
            set(date.formatLong() +
                    " --- Time: ${spent.formatHours()} / ${expected.formatHours()}" +
                    when {
                        spent < expected -> " --- Missing: ${(expected - spent).formatHours()}"
                        spent > expected -> " --- Extra: ${(spent - expected).formatHours()}"
                        spent == expected && expected != 0.0 -> " --- OK"
                        else -> "" // spent=expected=0
                    },
                getColor(expected, spent, date)
            )
        }
    }

    /**
     * Just for easier setters
     */
    fun set(label: String, color: Color? = null) = summary.apply { text = label; backgroundColor = color }
}

/**
 * Format LocalDate as LONG
 */
private fun LocalDate.formatLong() = format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG))